### ISSUE TITLE
Roadmap of upcoming major features

### ISSUE THREAD

**Issue by jrosdahl** [ccache/ccache#404]
### What's this?

This issue tracks major features that I (@jrosdahl) would like to see for ccache in the future. The tasks are roughly in a logical implementation order, but many can be worked on in parallel.

I will split the tasks to separate issues when time allows.

### Roadmap

- [x] 1. Introduce aggregated result files (#411)
- [x] 2. Switch from MD4 to BLAKE2b (#412)
- [x] 3. Switch from zlib to zstd (#413)
- [ ] 4. Design and implement a backend API and plugin framework (#414)
- [ ] 5. Refactor the current file system backend code to use the backend API.
  - Depends on task 4, and it would be a good idea to wait until tasks 1, 2 and 3 are done.
- [ ] 6. Implement a memcached backend.
  - Depends on tasks 4 and 5. See also #58.
- [ ] 7. Improve cache cleanup mechnism (#417)

**Comment by afbjorklund** [comment]
I am obviously in favor of 1, 2 and 3, since I suggested them. There are other checksums and other compressions, but those two are the best ones for the purpose IMHO. Should also mention that just like gzip, lz4f also has a "HC" option (like `--best`) that allows for recompressing for smaller size. The decompression speed is the same, it is only the compression speed that differs between the "levels".

**Comment by afbjorklund** [comment]
The main purpose of the memcached backend is to be able to _share_ cache over network between different computers. If you just want to put your cache in memory, you can use `/dev/shm` (ramdisk) already today. The choice of memcached for this is because it has been around for long and because it is fastest, but it does have some drawbacks like a 1M maxmum value size limit (worked around by splitting up data...)

**Comment by jrosdahl** [comment]
> I am obviously in favor of 1, 2 and 3, since I suggested them.

Great!

Just for the record, I have been considering single-result files [for a long time](https://www.mail-archive.com/ccache@lists.samba.org/msg01185.html). Bullet 2 and 3 are indeed inspired by your work and ideas, thanks!

> There are other checksums and other compressions, but those two are the best ones for the purpose IMHO

Yes, I have made a similar evaluation and came to the same conclusion (using LZ4 for a ccache-like tool at work).

> The main purpose of the memcached backend is to be able to share cache over network between different computers.

Of course.

**Comment by afbjorklund** [comment]
> > The main purpose of the memcached backend is to be able to share cache over network between different computers.
> 
> Of course.

I mention this, because the name ("memcached") can lead you to think it is about memory...
Still recommend using a local proxy though, but the actual server storage can be file-backed.

Tried to use a database for the ccache storage, but there was too much overhead in the setup.
So I think we need to change the architecture for that to work good, like client/server or a library ?

But the current code uses a lot of global variables and such, so it isn't exactly thread-safe.
We need to keep the cache running, either as a daemon or at least for the build (e.g. make)
This way the database setup only needs to be done once, and it can keep cache in memory.
The idea behind the library is that you could link it into your main build program, such as ninja.

Those approaches only help locally, though. Using memory means you are restricted to one host.
Could still be useful, if we continue with the idea of multiple caches - one local and one remote ?
Then the primary cache could use memory-mapped storage instead of just the file system like today.
And the secondary cache could replicate the cache to other servers, possible in the background.

But that's a **lot** of new stuff, much of it to be implemented into the main ccache distribution directly.

The best part of using existing software (such as [couchbase](https://www.couchbase.com/memcached-replacement)) was that it was available separately.

**Comment by afbjorklund** [comment]
Implementations:
1. Aggregated `.result`: #408 
2. BLAKE2b checksum: #406
3. LZ4f compression: #118

**Comment by pavsa** [comment]
Just a crazy question, wouldn’t it be more productive / less error prone to rewrite it to C++? No crazy template magic, just “C with classes” approach.  There’s lot of allocate/free on error patterns throughout the code, which would be unnecessary with RAII. Also a lot of built in hash tables, vectors, so on. Testing would be also easier.

**Comment by pavsa** [comment]
I would also add one more important thing when we change the hash algo:

Generate not hex string as a file name, but something that has the whole alphabet + numbers, possibly also uppercase chars. 

For example base36: “1Y2P0IJ32E8E7“ or base62 (with uppercase letters, but that might not work on windows). 

Reason: 
- shorter filenames
- more top level cache dirs / less waiting for stats lock during parallel builds (I get a lot of these with 32 cores building at the same time, as there are only 16 stats dirs)

**Comment by afbjorklund** [comment]
We could use base32, it has been done in other projects. Especially when moving to longer hashes.
My personal preference is base32hex*, but it doesn't really matter... But base36 just sounds "odd" ?

* it is the same thing, but without this weird human special treatment of the `'I'`and '`O'`characters.
as in: same algorithm, minus final lookup table. will probably implement this one, just to be "special".

**Comment by afbjorklund** [comment]
For stats, see #168 (and I'm sure the whole "stats" system could use a similar overhaul to the result)
@jrosdahl mentions that he has some secret special plans for improving cleanup, so maybe stats too ?

One long outstanding feature is to have stats per-build in addition to per-cache, mentioned in #262
So having multiple stats would be a good idea to plan, if designing a "stats" file replacement / addition.

**Comment by pavsa** [comment]
> 
> 
> We could use base32, it has been done in other projects. Especially when moving to longer hashes.
> My personal preference is [base32hex](https://tools.ietf.org/html/rfc2938)*, but it doesn't really matter... But base36 just sounds "odd" ?
> 
>     * it is the same thing, but without this weird human special treatment of the `'I'`and '`O'`characters.
>       as in: same algorithm, minus final lookup table. will probably implement this one, just to be "special".

You're right, base32 makes more sense. But the selected hash size in bits has to divisible by 5 so that it doesn't produce any padding. md4 is 128bit so it would produce padding. 160bit SHA-1 happens to be divisible by 5.

**Comment by afbjorklund** [comment]
@pavsa : for the current _binary_ key, the size is 160 bits (i.e. 128 bits MD4 + 32 bits size = 160 bits)
Some backends will just use this directly, but we could go with your suggestion for the file based one.

We will probably **not** use SHA-1 but BLAKE2b (for perfomance reasons mostly, and SHA-1 is old),
and it remains to be seen how many bits to use (the checksum itself is 512 bits). See #399 and #406.

**Comment by pavsa** [comment]
> We will probably **not** use SHA-1 but BLAKE2b (for perfomance reasons mostly, and SHA-1 is old),
> and it remains to be seen how many bits to use (the checksum itself is 512 bits). See #399 and #406.
Of course, larger hash can always be trimmed to a desirable size. Does it auto-detect CPU and run optimized code for it? How is it handled?

**Comment by afbjorklund** [comment]
Nah, we just assume `amd64` always (in terms of optimizations and code) and set a fixed hash size...

**Comment by pavsa** [comment]
But is faster though? 
I've done a simple benchmark of 1GB of data:
MD4 from openssl

Results:

MD4: time 1.815886s   563.9MB/s

blake2b (-O3 -march=corei7-avx from /sse dir): time 2.336596s   438.2MB/s
blake2b (-O3 -march=corei7-avx from /ref dir): time 2.643833s   387.3MB/s

**Comment by pavsa** [comment]
Btw, I did not switch to using stat64() for getting 64bit file sizes in fix for #382 
We are only storing 32bit sizes in 64bit integer so far.

**Comment by jrosdahl** [comment]
@afbjorklund & @pavsa: My intention with this issue was to track and discuss *which major tasks* to do and in *which order*, not to discuss specifics of those tasks. I would appreciate it if we all could make a habit to open new issues for tangent ideas and topics in the future.

Please move the different discussions to these issues:

- #412: Switch from MD4 to BLAKE2b and bump manifest on-disk format
- #415: Rewrite in C++? 
- #416: Change cache directory structure?

**Comment by jonesmz** [comment]
Another candidate for distribute object caching, in addition to memcached, would be the Ceph radios object storage api

**Comment by afbjorklund** [comment]
> Another candidate for distribute object caching, in addition to memcached, would be the Ceph radios object storage api

The workaround for now would be to use CephFS, but it could be interesting to compare that with using librados instead

**Comment by jonesmz** [comment]
I expect that the largest difference would come from avoiding the FUSE filesystem layer, with the second largest savings coming from skipping the metadata server.

**Comment by iveqy** [comment]
If I understand it correctly memcached is limited by RAM size while redis support cluster of redis instances. Even if memcached is faster/more mature it seems as it is easier to implement redis from a ccache perspective for large workloads. Or do I miss something?

**Comment by afbjorklund** [comment]
> If I understand it correctly memcached is limited by RAM size while redis support cluster of redis instances. Even if memcached is faster/more mature it seems as it is easier to implement redis from a ccache perspective for large workloads. Or do I miss something?

Please move discussion to a separate issue, thanks. (it _was_ easier, but the libmemcached implementation had optimizations)

**Comment by jrosdahl** [comment]
I'm closing this tracking issue now since it's obsolete. See also discussion in #414 and the [_Storage backends_ project](https://github.com/ccache/ccache/projects/1).
